tools/registry.py

```diff
@@ -0,0 +1,291 @@
+    self.build_targets.append(target)
+    return self
+
+  def add_test_targets(self, target):
+    if not target.startswith("@" + self.name):
+      target = "@" + self.name + target
+    self.test_targets.append(target)
+    return self
+
+  def dump(self, file):
+    json_dump(file, self.__dict__)
+
+  def from_json(self, file):
+    with open(file) as f:
+      self.__dict__ = json.load(f)
+
+
+class RegistryException(Exception):
+  """
+  Raised whenever something goes wrong with modifying the registry.
+  """
+
+
+class RegistryClient:
+  """A class to help create a Bazel registry."""
+
+  _MODULE_BAZEL = """
+module(
+    name = "{0}",
+    version = "{1}",
+    compatibility_level = {2},
+)
+""".strip()
+
+  def __init__(self, root):
+    self.root = pathlib.Path(root)
+
+  def contains(self, module_name, version=None):
+    """
+    Check if the registry contains a module or a specific version of a
+    module by verifying if the directory exists.
+    """
+    dir_path = self.root.joinpath("modules", module_name)
+    if version:
+      dir_path = dir_path.joinpath(version)
+    return dir_path.is_dir()
+
+  def init_module(self, module_name, maintainers, homepage):
+    """
+    Initialize a module, create the directory and metadata.json file.
+
+    Parameters
+    ----------
+    module_name : str
+        The module name
+    maintainers : list of maps of string -> string
+        The maintainer information, eg
+         [{"name": "John", "email": "john@guugoo.com"},
+          {"name": "Yun", "github": "meteorcloudy"}]
+    homepage : str
+        A URL to the project's homepage
+
+    """
+    p = self.root.joinpath("modules", module_name)
+    p.mkdir(parents=True, exist_ok=True)
+
+    # Create metadata.json file
+    metadata = {
+        "maintainers": maintainers,
+        "homepage": homepage,
+        "versions": [],
+        "yanked_versions": {},
+    }
+    json_dump(p.joinpath("metadata.json"), metadata)
+
+  def add(self, module, override=False):
+    """
+    Add a new module version, the module must be already initialized
+
+    Parameters
+    ----------
+    module : Module
+        A Module instance containing information of the module version to
+        be added
+    override : Whether to override existing module
+    """
+    # Check if the module version already exists
+    if self.contains(module.name, module.version):
+      if override:
+        log("Overridding module '%s' at version '%s'..." %
+            (module.name, module.version))
+        self.delete(module.name, module.version)
+      else:
+        raise RegistryException(
+            f"Version {module.version} for module {module.name} already exists.")
+
+    p = self.root.joinpath("modules", module.name, module.version)
+    p.mkdir()
+
+    # Create MODULE.bazel
+    module_dot_bazel = p.joinpath("MODULE.bazel")
+    if module.module_dot_bazel:
+      shutil.copy(module.module_dot_bazel, module_dot_bazel)
```

Comment:
Yeah, we could definitely add the check here along with other checks (like no override should be used), I added a TODO here since that could a slightly more complex change.